### PR TITLE
Basic2D PFCluster for each rechit seed

### PR DESCRIPTION
**Pull request by hatakeyamak**
#### PR description:

As discussed here: https://indico.cern.ch/event/926508/#9-review-of-pf-layer-clusterin
we often effectively create PFClusters directly from PFRecHit seed passing the seeds threshold for HF (e.g. at HLT). However, in the current code and configuration setup, we still go through the topo clustering step and iterative clustering step without doing actual clustering, and wasting sizable time.

With this new cluster producer from individual seeds, we get basically identical physics outputs (except tiny mismatches due to limited numerical precision in the current setup). And, this step becomes faster by a factor of 6-7 (check HF line below).

Using Run2018D/EphemeralHLTPhysics data:

Before this PR
```
FastReport   CPU time avg.      when run  Real time avg.      when run     Alloc. avg.      when run   Dealloc. avg.      when run  Modules
FastReport         1.3 ms         1.3 ms         1.3 ms         1.3 ms        +710 kB        +710 kB        -599 kB        -599 kB    hltParticleFlowClusterECALUncorrectedUnseeded
FastReport         5.6 ms         5.6 ms         5.6 ms         5.6 ms       +3659 kB       +3659 kB       -2886 kB       -2886 kB    hltParticleFlowClusterPSUnseeded
FastReport         4.9 ms         4.9 ms         4.9 ms         4.9 ms        +122 kB        +122 kB         -30 kB         -30 kB    hltParticleFlowClusterECALUnseeded
FastReport         3.0 ms         3.0 ms         3.0 ms         3.0 ms        +671 kB        +671 kB        -587 kB        -587 kB    hltParticleFlowClusterHBHE
FastReport         0.4 ms         0.4 ms         0.4 ms         0.4 ms        +274 kB        +274 kB        -197 kB        -197 kB    hltParticleFlowClusterHCAL
FastReport         5.2 ms         5.2 ms         5.2 ms         5.2 ms       +1670 kB       +1670 kB       -1247 kB       -1247 kB    hltParticleFlowClusterHF
```

With this PR
```
FastReport         1.3 ms         1.3 ms         1.3 ms         1.3 ms        +710 kB        +710 kB        -599 kB        -599 kB    hltParticleFlowClusterECALUncorrectedUnseeded
FastReport         5.6 ms         5.6 ms         5.6 ms         5.6 ms       +3659 kB       +3659 kB       -2886 kB       -2886 kB    hltParticleFlowClusterPSUnseeded
FastReport         4.9 ms         4.9 ms         4.9 ms         4.9 ms        +122 kB        +122 kB         -30 kB         -30 kB    hltParticleFlowClusterECALUnseeded
FastReport         3.0 ms         3.0 ms         3.0 ms         3.0 ms        +671 kB        +671 kB        -587 kB        -587 kB    hltParticleFlowClusterHBHE
FastReport         0.4 ms         0.4 ms         0.4 ms         0.4 ms        +274 kB        +274 kB        -197 kB        -197 kB    hltParticleFlowClusterHCAL
FastReport         0.7 ms         0.7 ms         0.7 ms         0.7 ms       +1154 kB       +1154 kB        -886 kB        -886 kB    hltParticleFlowClusterHF
```

#### PR validation:

Checked with with Run2018D/EphemeralHLTPhysics data 100 events that, HLT PF cluster and PF candidates stay the same within numerical precision.
Check with matrix tests: 10824.0, 11634.0, 23234.0, and 140.53 that they run fine.

#### if this PR is a backport please specify the original PR and why you need to backport that PR:

This is not backport.

@bendavid @jsalfeld @dimkarasav

### CODE REVIEW

**Comment by Martin-Grunewald** [comment]
Hi,
You only consider ```hltParticleFlowClusterHF``` but there are several more instances of ```"PFClusterProducer"``` within HLT menus:
```
cat OnLine_HLT_FULL.py | grep '"PFClusterProducer"'
process.hltParticleFlowClusterECALUncorrectedUnseeded = cms.EDProducer( "PFClusterProducer",
process.hltParticleFlowClusterPSUnseeded = cms.EDProducer( "PFClusterProducer",
process.hltParticleFlowClusterHBHE = cms.EDProducer( "PFClusterProducer",
process.hltParticleFlowClusterHF = cms.EDProducer( "PFClusterProducer",
process.hltParticleFlowClusterPSL1Seeded = cms.EDProducer( "PFClusterProducer",
process.hltParticleFlowClusterECALUncorrectedL1Seeded = cms.EDProducer( "PFClusterProducer",
process.hltParticleFlowClusterECALUncorrectedForMuonsMF = cms.EDProducer( "PFClusterProducer",
process.hltParticleFlowClusterPSForMuons = cms.EDProducer( "PFClusterProducer",
process.hltParticleFlowClusterHFForEgammaUnseeded = cms.EDProducer( "PFClusterProducer",
process.hltParticleFlowClusterECALUncorrectedForMuonsMFNoVtx = cms.EDProducer( "PFClusterProducer",
process.hltParticleFlowClusterPSForMuonsNoVtx = cms.EDProducer( "PFClusterProducer",
process.hltParticleFlowClusterECALUncorrectedForTkMuonsMF = cms.EDProducer( "PFClusterProducer",
process.hltParticleFlowClusterPSForTkMuons = cms.EDProducer( "PFClusterProducer,
process.hltParticleFlowClusterPSPPOnAA = cms.EDProducer( "PFClusterProducer",
process.hltParticleFlowClusterECALUncorrectedPPOnAA = cms.EDProducer( "PFClusterProducer",
```
Should some of these be considered as well? Eg, HF-like ```hltParticleFlowClusterHFForEgammaUnseeded```?

**Comment by hatakeyamak** [comment]
> Hi,
> You only consider `hltParticleFlowClusterHF` but there are several more instances of `"PFClusterProducer"` within HLT menus:
> 
> ```
> cat OnLine_HLT_FULL.py | grep '"PFClusterProducer"'
> process.hltParticleFlowClusterECALUncorrectedUnseeded = cms.EDProducer( "PFClusterProducer",
> process.hltParticleFlowClusterPSUnseeded = cms.EDProducer( "PFClusterProducer",
> process.hltParticleFlowClusterHBHE = cms.EDProducer( "PFClusterProducer",
> process.hltParticleFlowClusterHF = cms.EDProducer( "PFClusterProducer",
> process.hltParticleFlowClusterPSL1Seeded = cms.EDProducer( "PFClusterProducer",
> process.hltParticleFlowClusterECALUncorrectedL1Seeded = cms.EDProducer( "PFClusterProducer",
> process.hltParticleFlowClusterECALUncorrectedForMuonsMF = cms.EDProducer( "PFClusterProducer",
> process.hltParticleFlowClusterPSForMuons = cms.EDProducer( "PFClusterProducer",
> process.hltParticleFlowClusterHFForEgammaUnseeded = cms.EDProducer( "PFClusterProducer",
> process.hltParticleFlowClusterECALUncorrectedForMuonsMFNoVtx = cms.EDProducer( "PFClusterProducer",
> process.hltParticleFlowClusterPSForMuonsNoVtx = cms.EDProducer( "PFClusterProducer",
> process.hltParticleFlowClusterECALUncorrectedForTkMuonsMF = cms.EDProducer( "PFClusterProducer",
> process.hltParticleFlowClusterPSForTkMuons = cms.EDProducer( "PFClusterProducer,
> process.hltParticleFlowClusterPSPPOnAA = cms.EDProducer( "PFClusterProducer",
> process.hltParticleFlowClusterECALUncorrectedPPOnAA = cms.EDProducer( "PFClusterProducer",
> ```
> 
> Should some of these be considered as well? Eg, HF-like `hltParticleFlowClusterHFForEgammaUnseeded`?

Good point. Now hltParticleFlowClusterHFForEgammaUnseeded also adopts Basic2DClusterForEachSeed.

**Comment by Martin-Grunewald** [comment]
please test

**Comment by Martin-Grunewald** [comment]
+1

**Review by perrotta — COMMENTED** [review]
In a plugin you can even more conveniently get rid of the header file, and combine everything in the .cc one

File: RecoParticleFlow/PFClusterProducer/plugins/Basic2DClusterForEachSeed.cc [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+  seeds.reserve(hits.size());
+  for (unsigned int i = 0; i < hits.size(); ++i) {
+    if (!rechitMask[i] || !seedable[i])
+      continue;
+    seeds.emplace_back(i);
+  }
+
+  // loop over seeds and make clusters
+  reco::PFCluster cluster;
+  for (auto seed : seeds) {
```
Comment (fwyzard) [inline RecoParticleFlow/PFClusterProducer/plugins/Basic2DClusterForEachSeed.cc:21]: do you mean something like this instead ?
```suggestion
  for (unsigned int seed = 0; seed < hits.size(); ++i) {
```

Comment (perrotta) [reply RecoParticleFlow/PFClusterProducer/plugins/Basic2DClusterForEachSeed.cc:21]: Exactly

Comment (perrotta) [reply RecoParticleFlow/PFClusterProducer/plugins/Basic2DClusterForEachSeed.cc:21]: Well, without removing
```C++
  // loop over seeds and make clusters
  reco::PFCluster cluster;
```

Comment (fwyzard) [reply RecoParticleFlow/PFClusterProducer/plugins/Basic2DClusterForEachSeed.cc:21]: Added back :-)

Comment (hatakeyamak) [reply RecoParticleFlow/PFClusterProducer/plugins/Basic2DClusterForEachSeed.cc]: Thank you @perrotta @fwyzard .
I think I incorporated your suggestions and used this opportunity to further simplify it.

**Comment by perrotta** [comment]
please test

**Comment by perrotta** [comment]
+1
- Special cluster builder designed for use at HLT
- In the UnitTests outputs one can observe the reduction in cpu time for the affected HLT modules
- Jenkins tests pass and show no differences in reco outputs

**Comment by Martin-Grunewald** [comment]
+1

**Comment by qliphy** [comment]
+1

**Pull request merged**